treewide: fix and improve some `src`s

oxij opened the pull request:
This fixes some source-related build failures I noticed while building nixpkgs without using cache.nixos.org, and then adds some related improvements on top, to make these particular issues less likely in the future.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
- With
  - [X] `sandbox = true`
- [X] This should be a noop, since no hashes change.
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

oxij committed:
fetchurl: improve GNU Savannah mirrors

oxij committed:
treewide: use `mirror://savannah` when possible

... and improve some related `meta` URLs

oxij committed:
xmlsec: add a URL for older releases

drupol reviewed (commented):
How about replacing `sha256` with `hash` where it is relevant too then ?

oxij commented:
Well, I'd say it's out of scope of this, since this PR specifically on purpose does not touch any hashes.

drupol commented:
Just replacing those where it make sense, not updating any of them.

oxij committed:
treewide: use `hash = "sha256-.*"` in all touched files whenever possible

... but only when it does not require editing the hash itself, i.e.,  keeping
non-`base64` hashes as-is.

oxij commented:
Done for all `base64` hashes.

antonmosich commented on pkgs/development/python-modules/uncertainties/default.nix:
```diff
@@ -22,7 +22,8 @@ buildPythonPackage rec {
     repo = "uncertainties";
-    tag = version;
+    # tag = version; # will fail
+    rev = "1238d37c6ed87d018e696589716de488e7d9eca7";
```
In what way did `tag = version;` fail? It worked on my machine when I tried just now.

mmlb replied:
I just checked both tag and revision and both worked. I would drop this commit since it doesn't really make sense.

That README comment isnt about tags vs commits anyway.

oxij replied:
Re-checked this myself, and you are correct. So, done.

> README

It is about short revisions, but tag are also, essentially, short revisions, since GitHub did return 404 for me for this, I think it applies.

oxij committed:
fetchurl: update samba mirrors' URLs

oxij commented:
Okay, so apparently Samba mirrors changed their URL scheme again between me making that change and rebasing it on top of master, so that's fixed now.

Also, I checked most of the touched packages by `nix-build`ing their `.src`s with `hashed-mirrors = ` in my `nix.conf` (I checked all where I edited the hash, a bunch of Samba ones, and some random Savannah ones), and everything works.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2314

oxij commented:
Darwin checks are stuck. But maybe this is fine to merge anyway?

drupol commented:
Nope, all the tests are failing, I would not merge this at the moment.

oxij commented:
Well `xmlsec` test fails, but this PR is a noop, I fail to see how this could be at fault.

philiptaron commented on pkgs/applications/misc/devilspie2/default.nix:
```diff
@@ -15,8 +15,8 @@ stdenv.mkDerivation rec {
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/devilspie2/devilspie2-${version}.tar.xz";
-    sha256 = "Cp8erdKyKjGBY+QYAGXUlSIboaQ60gIepoZs0RgEJkA=";
+    url = "mirror://savannah/${pname}/${pname}-${version}.tar.xz";
```
Tested with `nix-build -A devilspie2.src --check` ✅

philiptaron commented on pkgs/applications/radio/klog/default.nix:
```diff
@@ -17,7 +17,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "https://download.savannah.nongnu.org/releases/klog/${pname}-${version}.tar.gz";
+    url = "mirror://savannah/${pname}/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A klog.src --check` ✅

drupol replied:
We usually avoid reusing pname, saw it too late.

oxij replied:
Why?

philiptaron commented on pkgs/build-support/fetchurl/mirrors.nix:
```diff
@@ -215,13 +215,19 @@
   samba = [
-    "https://www.samba.org/ftp/"
-    "http://www.samba.org/ftp/"
+    "https://download.samba.org/pub/"
+    "http://download.samba.org/pub/"
```
Verified by looking at the diff between these two.

philiptaron commented on pkgs/by-name/cu/cuyo/package.nix:
```diff
@@ -12,7 +12,7 @@ stdenv.mkDerivation {
 
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/cuyo/cuyo-2.1.0.tar.gz";
+    url = "mirror://savannah/cuyo/cuyo-2.1.0.tar.gz";
```
Tested with `nix-build -A cuyo.src --check` ✅

philiptaron commented on pkgs/by-name/fa/fastjar/package.nix:
```diff
@@ -13,7 +13,7 @@ stdenv.mkDerivation (finalAttrs: {
     pname = "fastjar-source";
     inherit (finalAttrs) version;
-    url = "https://download.savannah.gnu.org/releases/fastjar/fastjar-${finalAttrs.version}.tar.gz";
+    url = "mirror://savannah/fastjar/fastjar-${finalAttrs.version}.tar.gz";
```
Tested with `nix-build -A fastjar.src --check` ✅

philiptaron commented on pkgs/by-name/gu/guile-reader/package.nix:
```diff
@@ -14,7 +14,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "http://download.savannah.nongnu.org/releases/${pname}/${pname}-${version}.tar.gz";
+    url = "mirror://savannah/${pname}/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A guile-reader.src --check` ✅

philiptaron commented on pkgs/by-name/m1/m17n_db/package.nix:
```diff
@@ -12,8 +12,8 @@ stdenv.mkDerivation rec {
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/m17n/m17n-db-${version}.tar.gz";
-    sha256 = "sha256-SBJUo4CqnGbX9Ow6o3Kn4dL+R/w53252BEvUQBfEJKQ=";
+    url = "mirror://savannah/m17n/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A m17n_db.src --check` ✅

philiptaron commented on pkgs/by-name/mi/mi2ly/package.nix:
```diff
@@ -9,8 +9,8 @@ stdenv.mkDerivation rec {
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/mi2ly/mi2ly.${version}.tar.bz2";
-    sha256 = "sha256-lFbqH+syFaQDMbXfb+OUcWnyKnjfVz9yl7DbTTn7JKw=";
+    url = "mirror://savannah/${pname}/${pname}.${version}.tar.bz2";
```
Tested with `nix-build -A mi2ly.src --check` ✅

philiptaron commented on pkgs/by-name/sk/skribilo/package.nix:
```diff
@@ -29,7 +29,7 @@ stdenv.mkDerivation (finalAttrs: {
 
   src = fetchurl {
-    url = "http://download.savannah.nongnu.org/releases/skribilo/skribilo-${finalAttrs.version}.tar.gz";
+    url = "mirror://savannah/skribilo/skribilo-${finalAttrs.version}.tar.gz";
```
Tested with `nix-build -A skribilo.src --check` ✅

philiptaron commented on pkgs/by-name/st/storeBackup/package.nix:
```diff
@@ -36,7 +36,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/storebackup/storeBackup-${version}.tar.bz2";
+    url = "mirror://savannah/storebackup/storeBackup-${version}.tar.bz2";
```
Tested with `nix-build -A storeBackup.src --check` ✅

philiptaron commented on pkgs/development/python-modules/matplotlib/default.nix:
```diff
@@ -192,8 +192,8 @@ buildPythonPackage rec {
         src = fetchurl {
-          url = "https://download.savannah.gnu.org/releases/freetype/freetype-old/freetype-2.6.1.tar.gz";
-          sha256 = "sha256-Cjx9+9ptoej84pIy6OltmHq6u79x68jHVlnkEyw2cBQ=";
+          url = "mirror://savannah/freetype/freetype-old/freetype-2.6.1.tar.gz";
```
Tested with `nix-build -A python3Packages.matplotlib.tests.withOutdatedFreetype` ✅

philiptaron commented on pkgs/by-name/xm/xmlsec/package.nix:
```diff
@@ -21,8 +21,13 @@ lib.fix (
     src = fetchurl {
-      url = "https://www.aleksey.com/xmlsec/download/xmlsec1-${version}.tar.gz";
-      sha256 = "sha256-2C6TtpuKogWmFrYpF6JpMiv2Oj6q+zd1AU5hdSsgE+o=";
+      urls = [
```
Tested with `nix-build -A xmlsec.src --check` ✅

philiptaron commented on pkgs/by-name/xl/xlog/package.nix:
```diff
@@ -12,8 +12,8 @@ stdenv.mkDerivation rec {
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/xlog/${pname}-${version}.tar.gz";
-    sha256 = "sha256-NYC3LgoLXnJQURcZTc2xHOzOleotrWtOETMBgadf2qU=";
+    url = "mirror://savannah/${pname}/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A xlog.src --check` ✅

philiptaron commented on pkgs/tools/inputmethods/m17n-lib/default.nix:
```diff
@@ -11,7 +11,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/m17n/m17n-lib-${version}.tar.gz";
+    url = "mirror://savannah/m17n/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A m17n_lib.src --check` ✅

philiptaron commented on pkgs/tools/inputmethods/m17n-lib/otf.nix:
```diff
@@ -14,7 +14,7 @@ stdenv.mkDerivation rec {
 
   src = fetchurl {
-    url = "https://download.savannah.gnu.org/releases/m17n/${pname}-${version}.tar.gz";
+    url = "mirror://savannah/m17n/${pname}-${version}.tar.gz";
```
Tested with `nix-build -A libotf.src --check` ✅ 

(the file organization here is nuts)

philiptaron reviewed (approved):
I went through and validated every single change. This is good to merge.

oxij commented:
Thanks @philiptaron!
